File: solutions/530.minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int getMinimumDifference(TreeNode root) {
        int res = Integer.MAX_VALUE;
        int preVal = -1;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode cur = root;
        while(cur != null || !stack.isEmpty()){
            if(cur != null){
                stack.push(cur);
                cur = cur.left;
            }else{
                TreeNode node = stack.pop();
                if(preVal != -1 && res > Math.abs(node.val - preVal)){
                    res = Math.abs(node.val - preVal);
                }
                preVal = node.val;
                cur = node.right;
            }
        }
        return res;
    }
}

File: solutions/530.minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var getMinimumDifference = function(root) {
    if(root === null) return -1; 
    let preValue = undefined;
    let res = Infinity;
    helper(root);
    return res;
    
    function helper(node){
        if(node == null) return;
        helper(node.left);
        if(preValue!== undefined && res > Math.abs(node.val - preValue)){
            res = Math.abs(node.val - preValue);
        }
        preValue = node.val;
        helper(node.right);
    }
    
};